node_modules/gh-tests/lib/util.js

```diff
@@ -561,3 +561,85 @@ var parseIcalCalendar = module.exports.parseIcalCalendar = function(calendar, ca
+ *         {Object}     assertionFunction.actualData    Represents the "actual" data object you wish to run assertions on
+ *         {Object}     assertionFunction.assertions    An object keyed by assertion type (e.g., 'equals'), whose value is the assertion-specific data (usually "expected" data)
+ */
+var createAssertionFunction = module.exports.createAssertionFunction = function(customAssertionFunctions) {
+    customAssertionFunctions = customAssertionFunctions || {};
+
+    // Create our set of default generic assertion functions that are handy for most cases. See
+    // function summary for more information about what these functions do
+    var defaultAssertionFunctions = {
+        'equals': assert.deepEqual,
+        'exists': function(actual, exists) {
+            if (exists) {
+                assert.ok(actual);
+            } else {
+                assert.ok(!actual);
+            }
+        },
+        'has': function(actual, expectedFieldValues) {
+            if (_.isEmpty(expectedFieldValues)) {
+                assert.fail('Expected to assert the value of at least one field');
+            }
+
+            _.each(expectedFieldValues, function(value, key) {
+                assert.strictEqual(actual[key], value);
+            });
+        },
+        'missing': function(actual, missingKeys) {
+            if (!_.isArray(missingKeys) || _.isEmpty(missingKeys)) {
+                assert.fail('Expected a non-empty array of missing keys to test');
+            }
+
+            var invalidKeys = _.chain(actual)
+                .keys()
+                .intersection(missingKeys)
+                .value();
+            assert.ok(_.isEmpty(invalidKeys), util.format('Invalid keys existed on actual object: "%s"', invalidKeys.join('", "')));
+        }
+    };
+
+    // Overlay the default assertion functions onto the provided custom assertion functions
+    var assertionFunctions = _.defaults(customAssertionFunctions, defaultAssertionFunctions);
+
+    return function(actualData, assertions) {
+        if (_.isEmpty(assertions)) {
+            // There should always be at least one assertion
+            assert.fail('Expected at least one assertion to perform');
+        }
+
+        // Run each assertion that was specified by the consumer with the expected assertion data.
+        // The actual meaning of the data is dependent on the assertion implementation in
+        // `assertionFunctions`
+        _.each(assertions, function(expectedData, name) {
+            var assertionFunction = assertionFunctions[name];
+            if (!_.isFunction(assertionFunction)) {
+                // Fail hard if the consumer specified an assertion that doesn't have any
+                // functionality associated to it
+                assert.fail(util.format('Invalid assertion specified with name "%s"', name));
+            }
+
+            assertionFunction(actualData, expectedData);
+        });
+    };
+};
```

Comment:
This is what I propose for more advanced custom assertions of complex objects. It provides the ability to ensure it wasn't improperly used without having to duplicate the logic everywhere (e.g., didn't provide empty assertion or didn't specify non-existing assertion type), as well as some generic and customizable assertions.